File: Data Structures & Algorithms/count-number-of-islands/submission-7.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int R = grid.size();
        int C = grid[0].size();
        if (R == 0 || C == 0)
            return 0;
        int n = 0;

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (grid[i][j] == '1') {
                    dfs(i, j, grid);
                    cout << i << "\t" << j << "\n";
                    n++;
                }
            }
        }
        return n;
    }
    
    void dfs(int r, int c,vector<vector<char>>& grid) {
        if ((r >= grid.size()) || (c == grid[0].size()) || r < 0 || c <0
        || grid[r][c] == '0')
            return;
        grid[r][c] = '0';
        dfs(r+1, c, grid);
        dfs(r-1, c, grid);
        dfs(r, c-1, grid);
        dfs(r, c+1, grid);

    }
};

File: Data Structures & Algorithms/rotting-fruit/submission-7.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = deque()
        rows, cols = len(grid), len(grid[0])
        directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        fresh = 0
        time = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    fresh+=1
                elif grid[r][c] == 2:
                    q.append((r,c))
        while q and fresh > 0:
            for i in range(len(q)):
                row, col = q.popleft()
                for dr, dc in directions:
                    r, c = row+dr, col+dc
                    if r in range(rows) and c in range(cols) and grid[r][c] == 1:
                        fresh-=1
                        grid[r][c] = 2
                        q.append((r,c))
            time+=1 
        return time if fresh == 0 else -1
